### PR TITLE
Add support for DualSense trigger haptic feedback

### PR DESCRIPTION
**Pull request by mmmaisel**
Add support for DualSense trigger haptic feedback if SDL version is greater or equal to 2.0.16.
There are two supported effects, click and rigid which can be configured the the axis edit dialog
if a DualSense controller is connected. The feedback point follows the trigger point, i.e. the configured dead-zone.

In preparation of this change, remove some dead XML related code and clean-up the axis edit dialog.

### CODE REVIEW

**Comment by pktiuk** [comment]
> In preparation of this change, remove some dead XML related code and clean-up the axis edit dialog.

Could you open separate PR with these cleanups and removal of dead code? It would make repo history cleaner :)

**Comment by mmmaisel** [comment]
I extracted it into #504.

File: src/gamecontroller/gamecontroller.h [hunk 1]
```diff
@@ -81,6 +81,7 @@ class GameController : public InputDevice
     QHash<int, int> const &getDpadvalues();
 
     SDL_GameController *getController() const;
+    virtual int getControllerType() const override;
```
Comment (pktiuk) [inline src/gamecontroller/gamecontroller.h:84]: Why you use type `int` instead of `SDL_GameControllerType`?

File: src/gamecontroller/gamecontrollerset.h [hunk 2]
```diff
@@ -38,21 +38,17 @@ class GameControllerSet : public SetJoystick
...
-                                    QHash<int, SDL_GameControllerAxis> &axes,
-                                    QList<SDL_GameControllerButtonBind> &hatButtons); // GameControllerSetXml class
+    virtual void readConfig(QXmlStreamReader *xml);
 
   protected:
     void populateSticksDPad();
 
   public slots:
     virtual void reset();
+    void applyHapticFeedback();
```
Comment (pktiuk) [inline src/gamecontroller/gamecontrollerset.h:48]: I think a description of this method would be nice.

File: src/hapticfeedback.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+#include <SDL2/SDL_gamecontroller.h>
+#include <stdint.h>
+
+#include "hapticfeedbackmode.h"
+
+class HapticFeedback
+{
+  public:
+    explicit HapticFeedback(HapticFeedbackMode mode = HAPTIC_FEEDBACK_NONE, int strength = 0, int start = 0, int end = 0);
+
```
Comment (pktiuk) [inline src/hapticfeedback.h:29]: Why HapticFeedback is not `QObject`?

Comment (pktiuk) [reply src/hapticfeedback.h:29]: It is recommended, because than to this you can for example properly manage memory in your app [example](https://stackoverflow.com/a/2491811/11747535), QT [docs](https://doc.qt.io/qt-5/objecttrees.html)

File: src/gamecontroller/gamecontrollertrigger.cpp [hunk 4]
```diff
@@ -35,12 +38,27 @@ const GameControllerTrigger::ThrottleTypes GameControllerTrigger::DEFAULTTHROTTL
...
+#endif
+}
+
+GameControllerTrigger::~GameControllerTrigger()
+{
+    if (m_haptic_feedback)
+    {
+        delete m_haptic_feedback;
+        m_haptic_feedback = 0;
+    }
```
Comment (pktiuk) [inline src/gamecontroller/gamecontrollertrigger.cpp:61]: Followup to previous comment. In case of using `QObject` as a base for `HapticFeedback` this part would be skippable

File: src/gamecontroller/gamecontroller.h (hunk 1 above)
Comment (mmmaisel) [reply src/gamecontroller/gamecontroller.h:84]: SDL_GameControllerType is not available on SDL versions before 2.0.16.

File: src/gamecontroller/gamecontrollerset.h (hunk 2 above)
Comment (mmmaisel) [reply src/gamecontroller/gamecontrollerset.h:48]: Fixed in next patchset.

File: src/hapticfeedback.h (hunk 3 above)
Comment (mmmaisel) [reply src/hapticfeedback.h:29]: Fixed in next patchset.

File: src/gamecontroller/gamecontrollertrigger.cpp (hunk 4 above)
Comment (mmmaisel) [reply src/gamecontroller/gamecontrollertrigger.cpp:61]: Fixed in next patchset.

File: src/gamecontroller/gamecontroller.h (hunk 1 above)
Comment (pktiuk) [reply src/gamecontroller/gamecontroller.h:84]: What do you think about this hack?
```cpp
#if SDL_VERSION_ATLEAST(2, 0, 16)
#else
typedef enum
{
    SDL_CONTROLLER_TYPE_UNKNOWN = 0,
} SDL_GameControllerType;
#endif
```

**Comment by pktiuk** [comment]
I see this PR is targeted specifically for PS5 gamepad.  
```cpp
    if (parentSet->getInputDevice()->getControllerType() == SDL_GameControllerType::SDL_CONTROLLER_TYPE_PS5)
        m_haptic_feedback = new HapticFeedback(this);
```
Maybe it would be better to write a bit more generic code using SDL API-s used for haptics. (for example `SDL_HapticEffectSupported` for checking availability)

File: src/gamecontroller/gamecontroller.h (hunk 1 above)
Comment (mmmaisel) [reply src/gamecontroller/gamecontroller.h]: Done.

**Comment by mmmaisel** [comment]
> I see this PR is targeted specifically for PS5 gamepad.
> 
> ```c++
>     if (parentSet->getInputDevice()->getControllerType() == SDL_GameControllerType::SDL_CONTROLLER_TYPE_PS5)
>         m_haptic_feedback = new HapticFeedback(this);
> ```
> 
> Maybe it would be better to write a bit more generic code using SDL API-s used for haptics. (for example `SDL_HapticEffectSupported` for checking availability)

SDL_HapticEffect is a very old API from SDL 2.0.0 (2013) to control force feedback joysticks.
It can not control the PS5 controllers haptic triggers.

**Comment by pktiuk** [comment]
> It can not control the PS5 controllers haptic triggers.

You are right ( https://github.com/libsdl-org/SDL/issues/5125 )

File: src/gui/joycontrolstickeditdialog.ui [hunk 5]
```diff
@@ -7,7 +7,7 @@
     <x>0</x>
     <y>0</y>
     <width>702</width>
-    <height>570</height>
+    <height>568</height>
```
Comment (pktiuk) [inline src/gui/joycontrolstickeditdialog.ui:10]: Why this file is edited?

File: src/hapticfeedback.h [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+#include <QString>
+#include <SDL2/SDL_gamecontroller.h>
+#include <stdint.h>
+
+#include "hapticfeedbackmode.h"
+
+/**
+ * @brief Represents a single haptic feedback effect on a PS5 controller trigger.
+ */
+class HapticFeedback : QObject
```
Comment (pktiuk) [inline src/hapticfeedback.h:29]: If a class represents only a single haptic feedback for a single controller, then it shouldn't have such a generic name. It should be a bit more descriptive like `HapticTriggerPS5` if you think this class can be applicable only for this purpose or something like`HapticControllerPS5` if you think it can also have (or easily implement) other haptic-related features for PS5 Dualsense

File: src/hapticfeedback.h [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        EFFECT_LEFT_EN = 0x04,
+        EFFECT_RIGHT_EN = 0x08,
+        LEGACY_RUMBLE_LEFT_EN = 0x0001,
+        LEGACY_RUMBLE_RIGHT_EN = 0x0002,
+        MIC_LIGHT_EN = 0x0100,
+
+        MIC_LIGHT_OFF = 0x00,
+        MIC_LIGHT_SOLID = 0x01,
+        MIC_LIGHT_PULSE = 0x02
+    };
```
Comment (pktiuk) [inline src/hapticfeedback.h:55]: I think it would be good to provide link to source, where these values (and way of implementing this feature for PS5) can be found.

Comment (pktiuk) [reply src/hapticfeedback.h:55]: What does `MIC_LIGHT_` means? microphone light?

File: src/joyaxis.cpp [hunk 8]
```diff
@@ -295,6 +295,23 @@ int JoyAxis::calculateThrottledValue(int value)
...
+ * @brief Always returns nullptr for JoyAxis base objects.
+ */
+HapticFeedback *JoyAxis::getHapticFeedback() const { return nullptr; }
+
+/**
+ * @brief Changes the haptic feedback effect mode.
+ * @param[in] mode New haptic feedback effect mode.
+ */
+void JoyAxis::setHapticFeedbackMode(HapticFeedbackMode) {}
+
```
Comment (pktiuk) [inline src/joyaxis.cpp:314]: Just like above.  
A bit too general name. Methods' descriptions are good, but names like `hasHapticFeedback`suggest more generic haptic feedback (like classic rumble).

File: src/joyaxis.h [hunk 9]
```diff
@@ -41,7 +43,7 @@ class JoyAxis : public QObject
 
   public:
     explicit JoyAxis(int index, int originset, SetJoystick *parentSet, QObject *parent);
-    ~JoyAxis();
+    virtual ~JoyAxis();
```
Comment (pktiuk) [inline src/joyaxis.h:46]: Why did you change that?

File: src/gui/joycontrolstickeditdialog.ui (hunk 5 above)
Comment (mmmaisel) [reply src/gui/joycontrolstickeditdialog.ui]: Yes, this was accidentally committed. I removed it.

File: src/hapticfeedback.h (hunk 6 above)
Comment (mmmaisel) [reply src/hapticfeedback.h]: I renamed it to HapticTriggerPs5 so a HapticTrigger base class can be added later when there are more controllers with haptioc triggers.

File: src/hapticfeedback.h (hunk 7 above)
Comment (mmmaisel) [reply src/hapticfeedback.h]: I added a link to a good source and changed it to use the recommended messages.

MIC_LIGHT is the microphone light. I changed it as well.

File: src/joyaxis.cpp (hunk 8 above)
Comment (mmmaisel) [reply src/joyaxis.cpp]: I changed it to hasHapticTrigger.

File: src/joyaxis.h (hunk 9 above)
Comment (mmmaisel) [reply src/joyaxis.h:46]: I marked this a virtual for clarity. It was already virtual before by inheriting from QObject.

**Comment by pktiuk** [comment]
You did also a lot of nice cleanups in code :)  
Naming in `axiseditdialog` is much better now

**Review by pktiuk — APPROVED** [review]
